### DISCUSSION TITLE
Unable to dynamically link with libffi6 on Alma 8

### DISCUSSION THREAD

**Discussion by pmulgaonkar** [pyinstaller/pyinstaller#7395] [Help]
I have an issue with a simple python file compiled under Ubuntu 18 but running under Alma. Interestingly everything seems to run correct except for one subprocess call which invokes the system passwd command. The simple test file is:

```
#from cryptography.fernet import Fernet
import requests
import subprocess

out = subprocess.check_output("ls -l", shell=True)
print (out)
```
As in this version, with the `import requests` line uncommented, the execution looks like this:
```
[root@Alma8 u01]# ./test
b'total 70076\ndrwxr-xr-x. 3 raxak raxak      24 Dec 23 18:53 appliance4\n-rwxr-xr-x. 1 raxak raxak 9772432 Dec 23 18:44 Backer\n-rwxr-xr-x. 1 raxak raxak 9350080 Dec 23 18:44 Downloader\n-rwxr-xr-x. 1 raxak raxak     488 Jan 23 23:12 handshake.sh\n-rwxr-xr-x. 1 raxak raxak 9208416 Jan 24 18:21 Installer\n-rwxr-xr-x. 1 raxak raxak 6136584 Dec 23 18:44 Jogger\ndrwxr-xr-x. 2 raxak raxak       6 Dec 23 18:44 log\n-rwxr-xr-x. 1 raxak raxak 6539504 Dec 23 18:44 Recover\ndrwxr-xr-x. 3 raxak raxak      19 Dec 23 18:45 rp\n-rwxr-xr-x. 1 raxak raxak 6127448 Dec 23 18:44 Runner\n-rwxr-xr-x. 1 raxak raxak 9334904 Dec 23 18:44 Setup\n-rwxr-xr-x. 1 root  root  9127136 Jan 24 18:46 test\n-rwxr-xr-x. 1 raxak raxak 6139552 Dec 23 18:44 Upgrader\n'

passwd: error while loading shared libraries: libffi.so.6: failed to map segment from shared object
Traceback (most recent call last):
  File "test.py", line 7, in <module>
  File "subprocess.py", line 356, in check_output
  File "subprocess.py", line 438, in run
subprocess.CalledProcessError: Command 'passwd -S raxak | grep '^raxak'' returned non-zero exit status 1.
[108741] Failed to execute script 'test' due to unhandled exception!
```

With the `import requests` commented, everything runs fine... The output looks like 
```
[root@Alma8 u01]# ./test
b'total 64436\ndrwxr-xr-x. 3 raxak raxak      24 Dec 23 18:53 appliance4\n-rwxr-xr-x. 1 raxak raxak 9772432 Dec 23 18:44 Backer\n-rwxr-xr-x. 1 raxak raxak 9350080 Dec 23 18:44 Downloader\n-rwxr-xr-x. 1 raxak raxak     488 Jan 23 23:12 handshake.sh\n-rwxr-xr-x. 1 raxak raxak 9208416 Jan 24 18:21 Installer\n-rwxr-xr-x. 1 raxak raxak 6136584 Dec 23 18:44 Jogger\ndrwxr-xr-x. 2 raxak raxak       6 Dec 23 18:44 log\n-rwxr-xr-x. 1 raxak raxak 6539504 Dec 23 18:44 Recover\ndrwxr-xr-x. 3 raxak raxak      19 Dec 23 18:45 rp\n-rwxr-xr-x. 1 raxak raxak 6127448 Dec 23 18:44 Runner\n-rwxr-xr-x. 1 raxak raxak 9334904 Dec 23 18:44 Setup\n-rwxr-xr-x. 1 root  root  3354264 Jan 24 18:45 test\n-rwxr-xr-x. 1 raxak raxak 6139552 Dec 23 18:44 Upgrader\n'

b'raxak LK 2022-12-23 0 99999 7 -1 (Password locked.)\n'
[root@Alma8 u01]#
```
The cryptography library also seems to pull in the libffi.so.6. This file is a snippet from a longer piece of code that I created just to illustrate the issue. Interestingly in the full code, the actual calls to the requests library work fine. It is only the subprocess call to the external passwd command that fails.

I am using pyinstaller version 4.10, python3 version 3.6.9, and "compiling" the executable with

 `pyinstaller -F --clean "$app.py" --exclude-module pydoc`

Any help will be appreciated.

**Comment by bwoodsend** [comment]
This smells like another case of https://github.com/pyinstaller/pyinstaller/issues/6334. When you run your subprocess, `ls` receives the old version of `libffi.so` from your application instead of the system provided one it was intended to be linked against.

Whilst there are no universally correct ways of doing cross distro anything ABI related, generally if you're going down the *build on old docker images* route, you'll want to exclude pretty much any `.so` library that wasn't provided by a Python wheel.

**Comment by pmulgaonkar** [comment]
So in my particular case do I just put --exclude libffi in the 
pyinstaller invocation? That does not do anything. I tried it.

or do I need to put --exclude libffi.so?

I am confused as to what exactly you are recommending. I looked at the 
6334 issue, but couldn't map it to my case exactly.


--p

**Reply by rokm** [reply]
No, `--exclude` works only for python modules. Since you're using onefile build, the only way to exclude a shared library from collection is editing `Analysis.binaries` in the spec file and building from spec file instead of from .py file.

But the proper way to handle such situations is to properly set `LD_LIBRARY_PATH` in your subprocess' environment - i.e., restoring it from `LD_LIBRARY_PATH_ORIG` (if available) or just un-setting it (if original is unavailable, which means that it was not explicitly set before PyInstaller did its modifications). See [the documentation](https://pyinstaller.org/en/stable/runtime-information.html#ld-library-path-libpath-considerations). There's also examples in discussion under #6334.

**Comment by pmulgaonkar** [comment]
Thanks. That was it. I had done this in other parts of the code but 
missed this set of occurrences.

Appreciate it.

--p
